### PR TITLE
Add tests for blog app

### LINKED ISSUE #16: Create tests
- [x] Read about tests [here](http://www.drdobbs.com/testing/unit-testing-with-python/240165163)
- [x] Also there's a good link about test's in django docs [here](https://docs.djangoproject.com/en/1.10/topics/testing/)
- [x] Create tests for models
- [x] Create tests for views
- [x] Integrate your app with Travis CI and run test at Travis' script

As an example you can use [this](https://github.com/kpi-web-guild/django-girls-blog-DrEdi/blob/master/vagga.yaml), but try to do it by yourself :new_moon_with_face:

### PR DESCRIPTION
**Pull request by PetrushynskyiOleksii**
**Close #16**

### CODE REVIEW

File: .travis.yml [hunk 1]
```diff
@@ -1,5 +1,9 @@
 language: python
 python: 3.6
 cache: pip
-before_install: pip install -U pre-commit
-script: pre-commit run --all-files
+before_install:
+    - pip install -U pre-commit
```
Comment (webknjaz) [inline .travis.yml]: Dedent these lines. Sequences do not require indentation.

File: blog/tests/__init__.py [hunk 2]
```diff
@@ -0,0 +1 @@
+"""Package of test for blog app."""
```
Comment (webknjaz) [inline blog/tests/__init__.py]: Just "test suite", "package" is an implementation detail.

File: blog/tests/test_models.py [hunk 3]
```diff
@@ -0,0 +1,30 @@
...
+                                             text='Text Test 123 !@#')
+
+    def tearDown(self):
+        """Delete test data."""
+        del self.user
+        del self.test_post
+
+    def test_post_publish(self):
+        """Post published successfully."""
+        self.test_post.publish()
```
Comment (webknjaz) [inline blog/tests/test_models.py]: What are you testing here? 0_O

File: blog/tests/test_models.py [hunk 4]
```diff
@@ -0,0 +1,30 @@
...
+                                             title='Title Test',
+                                             text='Text Test 123 !@#')
+
+    def tearDown(self):
+        """Delete test data."""
+        del self.user
+        del self.test_post
+
+    def test_post_publish(self):
+        """Post published successfully."""
```
Comment (webknjaz) [inline blog/tests/test_models.py]: How do you know that it's actually published?

File: blog/tests/test_models.py [hunk 5]
```diff
@@ -0,0 +1,30 @@
...
+from django.test import TestCase
+
+from blog.models import Post
+
+
+class PostModelTest(TestCase):
+    """Tests for post model."""
+
+    def setUp(self):
+        """Init test data."""
```
Comment (webknjaz) [inline blog/tests/test_models.py]: "Pre-populate"

File: blog/tests/test_models.py [hunk 6]
```diff
@@ -0,0 +1,30 @@
...
+        self.user = User.objects.create(username='UserTest')
+        self.test_post = Post.objects.create(author=self.user,
+                                             title='Title Test',
+                                             text='Text Test 123 !@#')
+
+    def tearDown(self):
+        """Delete test data."""
```
Comment (webknjaz) [inline blog/tests/test_models.py]: Clean-up

File: blog/tests/test_views.py [hunk 7]
```diff
@@ -0,0 +1,98 @@
...
+        self.client = Client()
+        self.user = User.objects.create(username=self.USERNAME,
+                                        email='testuser@test.com',
+                                        is_superuser=True,
+                                        is_staff=True,
+                                        is_active=True)
+        self.user.set_password(self.PASSWORD)
+        self.user.save()
+        self.tz = timezone.get_current_timezone()
+        self.post1 = Post.objects.create(author=self.user,
```
Comment (webknjaz) [inline blog/tests/test_views.py]: Could you please use more comprehensive names for posts?

File: blog/tests/test_views.py [hunk 8]
```diff
@@ -0,0 +1,98 @@
...
+
+    def setUp(self):
+        """Init test data."""
+        self.client = Client()
+        self.user = User.objects.create(username=self.USERNAME,
+                                        email='testuser@test.com',
+                                        is_superuser=True,
+                                        is_staff=True,
+                                        is_active=True)
+        self.user.set_password(self.PASSWORD)
```
Comment (webknjaz) [inline blog/tests/test_views.py]: look for `Model.objects.create_superuser`, it should be possible to do all steps with one call.

File: blog/tests/test_views.py [hunk 9]
```diff
@@ -0,0 +1,98 @@
...
+        del self.tz
+        del self.client
+        del self.user
+
+    def test_post_list(self):
+        """Test list of posts."""
+        response = self.client.get(reverse('post_list'))
+        self.assertEqual(response.status_code, 200)
+        self.assertEqual(len(response.context['posts']), 3)
+        self.assertNotContains(response, self.future_post)
```
Comment (webknjaz) [inline blog/tests/test_views.py]: What will happen if you run this test in 2223rd year?

File: blog/tests/test_views.py [hunk 10]
```diff
@@ -0,0 +1,98 @@
...
+        self.assertTemplateUsed(response, 'blog/post_list.html', 'blog/base.html')
+
+    def test_sorting_post_list(self):
+        """Test post sorting."""
+        response = self.client.get(reverse('post_list'), {'order_by': 'published_date'})
```
Comment (webknjaz) [inline blog/tests/test_views.py]: Are they all published by default?

Comment (webknjaz) [reply blog/tests/test_views.py]: Do you test case for unpublished post?

File: blog/tests/test_views.py [hunk 11]
```diff
@@ -0,0 +1,98 @@
...
+
+    def test_post_new_view(self):
+        """Test view for add post."""
+        authorization = self.client.login(username=self.USERNAME, password=self.PASSWORD)
+        self.assertTrue(authorization)
+        response = self.client.post(reverse('post_new'),
+                                    {'author': self.user, 'title': 'Test', 'text': 'superText', },
+                                    follow=True)
+        self.assertEqual(200, response.status_code)
+        self.post = Post.objects.get(author=self.user, title='Test', text='superText')
```
Comment (webknjaz) [inline blog/tests/test_views.py]: Woah! Do you use it outside of the test 0_O? Where's cleanup for it?

File: blog/tests/test_views.py [hunk 12]
```diff
@@ -0,0 +1,98 @@
...
+        self.assertContains(response, self.post)
+
+    def test_post_edit(self):
```
Comment (webknjaz) [inline blog/tests/test_views.py]: You don't test editing at all.

**Review by webknjaz — CHANGES_REQUESTED** [review]

File: .travis.yml [hunk 13]
```diff
@@ -1,5 +1,12 @@
 language: python
 python: 3.6
 cache: pip
-before_install: pip install -U pre-commit
-script: pre-commit run --all-files
+
+
+before_install:
+  - pip install -U pre-commit
```
Comment (webknjaz) [inline .travis.yml]: still indented :(

File: blog/tests/test_views.py [hunk 14]
```diff
@@ -0,0 +1,108 @@
...
+    USERNAME = 'testuser'
+    PASSWORD = 'password'
+    EMAIL = 'email@test.com'
+
+    def setUp(self):
+        """Pre-populate test data."""
+        self.client = Client()
+        self.user = User.objects.create_superuser(username=self.USERNAME,
+                                                  password=self.PASSWORD,
+                                                  email=self.EMAIL,)
```
Comment (webknjaz) [inline blog/tests/test_views.py]: put the closing paren on the next line: trailing commas are there to allow cleaner diffs, if you put parentheses there it becomes useless.

File: blog/tests/test_views.py [hunk 15]
```diff
@@ -0,0 +1,108 @@
...
+        del self.client
+        del self.user
+
+    def test_post_list(self):
+        """Test list of posts."""
+        response = self.client.get(reverse('post_list'))
+        self.assertEqual(response.status_code, 200)
+        self.assertEqual(len(response.context['posts']), 3)
+        self.assertTemplateUsed(response, 'blog/post_list.html', 'blog/base.html')
+        with patch('django.utils.timezone.now', lambda: datetime(day=1, month=1, year=2000, tzinfo=self.tz)):
```
Comment (webknjaz) [inline blog/tests/test_views.py]: WAT

File: blog/tests/test_views.py [hunk 16]
```diff
@@ -0,0 +1,108 @@
...
+        del self.second_post
+        del self.third_post
+        del self.future_post
+        del self.tz
+        del self.client
+        del self.user
+
+    def test_post_list(self):
+        """Test list of posts."""
+        response = self.client.get(reverse('post_list'))
```
Comment (webknjaz) [inline blog/tests/test_views.py]: So what exactly happens in year 2223?

File: blog/tests/test_views.py [hunk 17]
```diff
@@ -0,0 +1,108 @@
...
+                                     'title': 'Test New Post',
+                                     'text': 'Test Text',
+                                     },
+                                    follow=True)
+        self.assertEqual(200, response.status_code)
+        self.assertContains(response, 'Test New Post')
+
+    def test_post_edit(self):
+        """Test view for edit post."""
+        authorization = self.client.login(username=self.USERNAME, password=self.PASSWORD)
```
Comment (webknjaz) [inline blog/tests/test_views.py:96]: Test access without auth as well.

**Review by webknjaz — CHANGES_REQUESTED** [review]

File: blog/tests/test_views.py (hunk 17 above)
Comment (PetrushynskyiOleksii) [reply blog/tests/test_views.py:96]:
```
ValueError: Cannot assign "<SimpleLazyObject: <django.contrib.auth.models.AnonymousUser object at 0x7f37a9346c88>>": "Post.author" must be a "User" instance.
```

Comment (webknjaz) [reply blog/tests/test_views.py:96]: ?

**Review by webknjaz — CHANGES_REQUESTED** [review]
WAT

File: blog/tests/test_views.py (hunk 17 above)
Comment (PetrushynskyiOleksii) [reply blog/tests/test_views.py:96]: if I remove the authorization line, it will knock this error

File: blog/tests/test_views.py [hunk 18]
```diff
@@ -53,11 +54,10 @@ def tearDown(self):
 
     def test_post_list(self):
         """Test list of posts."""
-        response = self.client.get(reverse('post_list'))
-        self.assertEqual(response.status_code, 200)
-        self.assertEqual(len(response.context['posts']), 3)
-        self.assertTemplateUsed(response, 'blog/post_list.html', 'blog/base.html')
```
Comment (webknjaz) [inline blog/tests/test_views.py]: Why did you remove template check?

File: blog/tests/test_views.py [hunk 19]
```diff
@@ -0,0 +1,108 @@
...
+    def test_post_detail_view(self):
+        """Test detail page of specific post."""
+        response = self.client.get(reverse('post_detail', kwargs={'pk': 9999}))
+        self.assertEqual(404, response.status_code)
+        response = self.client.get(reverse('post_detail', kwargs={'pk': self.first_post.pk}))
+        self.assertEqual(200, response.status_code)
+
+    def test_post_new_view(self):
+        """Test view for add post."""
+        authorization = self.client.login(username=self.USERNAME, password=self.PASSWORD)
```
Comment (webknjaz) [inline blog/tests/test_views.py:83]: It's good to try posting new article before doing auth.

File: blog/tests/test_views.py (hunk 17 above)
Comment (webknjaz) [reply blog/tests/test_views.py:96]: Which means you didn't implement raising HTTP 401 in case of unauthorized user.
Also, it's good to try editing another user's post.

**Review by webknjaz — CHANGES_REQUESTED** [review]

**Review by webknjaz — APPROVED** [review]
